go/logictest/runner.go

```diff
@@ -246,11 +251,45 @@ func runTestFile(harness Harness, file string) {
 	}
 }
 
+type R struct {
+	schema string
+	results []string
+	cont bool
+	err error
+}
+
 // Executes a single record and returns whether execution of records should continue
 func executeRecord(harness Harness, record *parser.Record) (schema string, results []string, cont bool, err error) {
-	startTime = time.Now()
 	currRecord = record
+	startTime = time.Now()
+
+	ctx, cancel := context.WithTimeout(context.Background(), timeout)
+	defer cancel()
+
+	rc := make(chan *R, 1)
+	go func() {
+		schema, results, cont, err := execute(ctx, harness, record)
+		rc <- &R{
+			schema: schema,
+			results: results,
+			cont: cont,
+			err: err,
+		}
+	}()
+
+	select {
+	case res := <-rc:
+		return res.schema, res.results, res.cont, res.err
+	case <-ctx.Done():
+		logTimeout()
```

Comment:
You still have a small race here. Execute could already have logged its result (or be in the middle of it) before the timeout triggers, which means you could still be logging results for this test more than once.

Reply:
Fixing this issue is annoying. You need to define a locking construct and give it to the context, then use it to gate access to printing methods. Something like this:

```
type LoggingLock struct {
  mux sync.Mutex
  logged bool
}

ctx = timeoutContext.WithValue('lock', &LoggingLock)
```

Then whenever you are about to log from anywhere, you need to:

```
lock = ctx.Value('lock').(*LoggingLock)
lock.mux.Lock()
defer lock.mux.Unlock()
if lock.logged {
  return
}
logSomething()
lock.logged = true
```